src/main/java/com/google/sps/servlets/UploadServlet.java

```diff
@@ -0,0 +1,140 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+ 
+package com.google.sps.servlets;
+ 
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.MultipartConfig;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.Part;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.nio.file.Files;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import java.nio.file.Paths;
+import com.google.common.collect.ImmutableList;
+
+/** Servlet for uploading files. */
+@WebServlet("/upload")
+@MultipartConfig
+public class UploadServlet extends HttpServlet {
+
+  // The ID of your GCP project
+  private static final String PROJECT_ID = "step-2020-johndallard";
+
+  // The ID of your GCS bucket
+  private static final String BUCKET_NAME = "embedding-visualizer-bucket";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+    //  Get all the Dataset Entities
+    Query query = new Query("MetaData").addSort("timestamp", SortDirection.ASCENDING);
+
+    /* Get current user logged in to webapp */
+    UserService userService = UserServiceFactory.getUserService();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    /*  Get metadata (for comparison and eventual storage) */
+    String userEmail = userService.getCurrentUser().getEmail();
+    String datasetName = request.getParameter("dataset-namer");
+    long timestamp = System.currentTimeMillis();
+
+
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("user-email");
+      String dataset = (String) entity.getProperty("dataset-name");
+
+      if (email.equals(userEmail)) {
+        if (dataset.equals(datasetName)) {
+          // This dataset already exists for the user
+          response.sendRedirect("/already_exists.html");
```

Comment:
What do you mean by return? I guess I'm confused by that

Reply:
I mean
"return;"
after the response.sendRedirect line.
So that it doesn't process further?

Reply:
Ohhh I just assumed with response.sendRedirect() it won't process anything further. At least that's how its been working, because it doesn't end up uploading the images or creating new subdirectories in the bucket

Reply:
But I'll just add it then

Reply:
I don't think Java works that way. You need an explicit return statement to quit processing in the middle of a function.